snapcraft/tests/test_commands_stage.py

```diff
@@ -0,0 +1,149 @@
+        self.assertTrue(os.path.exists(parts[0]['state_file']),
+                        'Expected a state file for the build0 part')
+
+        with open(parts[0]['state_file']) as sf:
+            state = sf.readlines()
+        self.assertEqual(len(state), 1, 'Expected only one line in the state '
+                         'file for the build0 part')
+        self.assertEqual(state[0], 'stage', "Expected the state file for "
+                         "build0 to be 'stage'")
+
+    def test_stage_one_part_only_from_3(self):
+        fake_logger = fixtures.FakeLogger(level=logging.ERROR)
+        self.useFixture(fake_logger)
+        parts = self.make_snapcraft_yaml(n=3)
+
+        stage.main(['stage1', ])
+
+        self.assertTrue(os.path.exists(common.get_stagedir()),
+                        'Expected a stage directory')
+        self.assertTrue(os.path.exists(common.get_partsdir()),
+                        'Expected a parts directory')
+        self.assertTrue(os.path.exists(parts[1]['part_dir']),
+                        'Expected a part directory for the stage1 part')
+        self.assertTrue(os.path.exists(parts[1]['state_file']),
+                        'Expected a state file for the stage1 part')
+
+        with open(parts[1]['state_file']) as sf:
+            state = sf.readlines()
+        self.assertEqual(len(state), 1, 'Expected only one line in the state '
+                         'file for the stage1 part')
+        self.assertEqual(state[0], 'stage', "Expected the state file for "
+                         " stage1 to be 'stage'")
+
+        for i in [0, 2]:
+            self.assertFalse(os.path.exists(parts[i]['part_dir']),
+                             'Pulled wrong part')
+            self.assertFalse(os.path.exists(parts[i]['state_file']),
+                             'Expected for only to be a state file for build1')
+
+    def test_stage_ran_twice_is_a_noop(self):
+        fake_logger = fixtures.FakeLogger(level=logging.ERROR)
+        self.useFixture(fake_logger)
+        parts = self.make_snapcraft_yaml()
+
+        stage.main()
+
+        self.assertTrue(os.path.exists(common.get_stagedir()),
+                        'Expected a stage directory')
+        self.assertTrue(os.path.exists(common.get_partsdir()),
+                        'Expected a parts directory')
+        self.assertTrue(os.path.exists(parts[0]['part_dir']),
+                        'Expected a part directory for the build0 part')
+        self.assertTrue(os.path.exists(parts[0]['state_file']),
+                        'Expected a state file for the build0 part')
+
+        with open(parts[0]['state_file']) as sf:
+            state = sf.readlines()
+        self.assertEqual(len(state), 1, 'Expected only one line in the state '
+                         'file for the build0 part')
+        self.assertEqual(state[0], 'stage', "Expected the state file for "
+                         "build0 to be 'stage'")
+
+        stage.main()
```

Comment:
What is the assertion here?
Maybe we can refactor the step functions to be:
def step(...):
   if not should_step_run(): 
       return
    _step(...)

Put mocks in those _step methods, and check that they were not called.

Or maybe checking that the logger has the skip messages.